prime-router/docs/proposals/0017-serializer-error-codes.md

```diff
@@ -0,0 +1,243 @@
+the front-end without any cleaning or helpful metadata. These types of errors are not helpful to the end-user 
+because they are hard to read and often times don't give the user enough information to fix the issue.
+
+## Solution
+ReportStream should catch exceptions thrown by third-party libraries and wrap them in a specific error code, 
+combined with helpful metadata,
+that the front-end can map to a nice, helpful message for the end-user. For example, ReportStream can have 
+an error code for when an exception gets thrown when processing a date field, say `INVALID_HL7_GENERAL_PARSE`, and part of
+the JSON that houses the error code would also have `Type=DATE` and `Field=PID-29`.
+With this information, front-end would then know to show the user a helpful message regarding what date formats are 
+acceptable, instead of showing a stack-trace.
+
+In addition to third-party library exceptions, ReportStream can also run into issues when trying to load parsed
+values into Objects (Element). ReportStream should handle this case gracefully as well. This will be elaborated on down below and
+the process of loading a parsed value into an Object will be referred to as "normalization," to match the language 
+used in the code. One important thing to note here right away is that when ReportStream tries to load the parsed value as an element,
+it has additional information that it doesn't in the external library case, and that is "what exactly is wrong?", 
+along with the actual parsed value. This allows for the possibility of a more detailed error code, like `INVALID_HL7_DATE_SCHEMA_TYPE`
+instead of `INVALID_HL7_GENERAL_PARSE`.
+
+# Design
+
+Please note, the following design will be specific to HL7 but should be expandable enough for other formats as well.
+
+## Action Messages
+
+ReportStream defines various error/warning messages in `ActionMessages` which eventually get sent to the front-end
+as JSON. These messages can be expanded to contain a new field, called `errorCode`, along with any other extra metadata
+that would be helpful for the front-end, such as the specific field in question, its type, or its parsed value.
+The front-end can then use the errorCode and other metadata in the messages to calculate and display user-friendly messages.
+
+The following is the definition of the new action message that should be added. Its usage will be discussed in the
+sections below.
+
+```kotlin
+class FieldProcessingMessage(
+    // Name of the HL7 field
+    fieldMapping: String = "",
+    // the ReportStream Element type associated with fieldMapping 
+    // for the particular schema associated
+    fieldType: String = "",
+    // the raw value that was parsed out by an external library (HAPI)
+    fieldValue: String = "",
+    // Back-end error message. front-end can choose to display 
+    // this or a more friendly message using the other fields
+    override val message: String,
```

Comment:
This is not guaranteed to be "nice". Meaning, in the case of most (all?) HAPI errors this could potentially be an ugly stack trace. Are you suggesting the back-end to more processing and ALWAYS come up with a nice message?

Reply:
The idea would be that the front-end would be able to calculate their message based on all the fields in FieldProcessingMessage.

Reply:
No, I'm not suggesting that this always be a user-friendly string, just in the case of when we'd want to include formatted messages in the response.  But it's simple to handle this conversion on the front end 👍